Apps/W1/Sustainability/app/src/Journal/SustainabilityJnlLine.Table.al

```diff
@@ -334,15 +334,30 @@ table 6214 "Sustainability Jnl. Line"
         Validate("Document No.", SustainabilityJournalMgt.GetDocumentNo(IsPreviousLineValid, SustainabilityJnlBatch, PreviousLine."Document No.", "Posting Date"));
     end;
 
-    local procedure GetDefaultDimensionsFromAccount()
+    procedure CreateDimFromDefaultDim(FieldNo: Integer)
```

Comment:
Looking into the code path, looks like we are indeed lacking the capability like the proposed event `OnAfterInitDefaultDimensionSources`.

However, these two procedures look generic enough to move to `DimensionManagement.Codeunit.al`?

Or maybe even an interface instead?

Reply:
To be honest.
I don't see how we should move tese procedures to Dimension Management.
First of all. 
1. If we want to move them to the Dimension Management, then we would have to add procedures with variants to make them generic. -> If you convert a variant back to a record then all global variables are lost in the future. Global variable is state is sometimes very important e.g. for the SalesLine.
2. The procedures are very slim already and everything can be be abstracted is alreday reused from the DimensionManagement. The only remaining part depends on the `Rec` of the table.
3. I think it would be very useful to be able to define generic interfaces with a generic Type that tables can implement (e.g. `table 50000 "PTE Journal Line" impelments "IDefaultDimension"`), but that is currently not possible. 
4. Moving forward: Moving the DimensionManagement to the Business Foundation Layer , I think that it's easier to have generic helper procedures in the DimensionManagement Codeunit and only use this part in specific areas. Otherwise we will get several AreaSpecific Event Subscribers for the area/table specific calls. For this specific example we then would have a "SustainabilityDimMgt" which would subscribe to a generic `OnAfterInitDefaultDimensionSources`. We can discuss this maybe in a larger group to get this part of the BaseApp refactored, but I would prefer if we can get the current events and procedures in the Sustainability App at first.

Reply:
I would really like the idea of an interface, but for this we would need to be able to delcare event publishers in interfaces.
https://experience.dynamics.com/ideas/idea/?ideaid=98168f2d-c522-ec11-b76a-0003ff45de6f
Or maybe I don't understand what you expect on the interface.

This is what I would expect.

```al
interface "IJournalDefaultDimension"
{
    procedure CreateDim(DefaultDimSource: List of [Dictionary of [Integer, Code[20]]])

    procedure CreateDim(DefaultDimSource: List of [Dictionary of [Integer, Code[20]]])

    [IntegrationEvent(false, false)]
    local procedure <T>OnAfterInitDefaultDimensionSources(var Rec: Record T; var DefaultDimSource: List of [Dictionary of [Integer, Code[20]]]; FieldNo: Integer)
}
```
We would need a generic type for this kind of interfaces.

```al
table 6214 "Sustainability Jnl. Line" implements IJournalDefaultDimension
{
...
    procedure CreateDimFromDefaultDim(FieldNo: Integer)
    var
        DefaultDimSource: List of [Dictionary of [Integer, Code[20]]];
    begin
        InitDefaultDimensionSources(DefaultDimSource, FieldNo);
        CreateDim(DefaultDimSource);
    end;

    local procedure InitDefaultDimensionSources(var DefaultDimSource: List of [Dictionary of [Integer, Code[20]]]; FieldNo: Integer)
    begin
        DimMgt.AddDimSource(DefaultDimSource, Database::"Sustainability Account", "Account No.", FieldNo = Rec.FieldNo("Account No."));
        OnAfterInitDefaultDimensionSources(Rec, DefaultDimSource, FieldNo);
    end;

    procedure CreateDim(DefaultDimSource: List of [Dictionary of [Integer, Code[20]]])
    begin
        "Shortcut Dimension 1 Code" := '';
        "Shortcut Dimension 2 Code" := '';
        "Dimension Set ID" := DimMgt.GetRecDefaultDimID(Rec, CurrFieldNo, DefaultDimSource, "Source Code", "Shortcut Dimension 1 Code", "Shortcut Dimension 2 Code", 0, 0);
    end;

    [IntegrationEvent(false, false)]
    local procedure OnAfterInitDefaultDimensionSources(var SustainabilityJnlLine: Record "Sustainability Jnl. Line"; var DefaultDimSource: List of [Dictionary of [Integer, Code[20]]]; FieldNo: Integer)
    begin
    end;
}
```

Reply:
Resolving conversation to process PR. Haroan and team have agreed to go with this solution. @haoranpb, please finish discussion despite the conversation being resolved.

Reply:
@pri-kise We have discussed this internally and agreed to move forward with the PR, thanks Jesper for merging it.

Just want to drop my thought process here:

For Variants and Global variables, personally, I think the use of global variables should be justified on a case-by-case basis, it can become really hard to understand when/where a global variable is set in places like `codeunit 80 "Sales-Post"`.

Agree on that the procedures here are fairly slim already, and it should be moved when we get to the refactoring of the DimensionManagement.